### PR TITLE
Write hooks with `{'primary_key': , 'timestamp':}`, also fixes r.now determinism problems

### PR DESCRIPTION
**Pull request by srh**
- [x] I have read and agreed to the RethinkDB Contributor License Agreement http://rethinkdb.com/community/cla/

### Description

This implements #6267, which is a ReQL proposal.  Has it resolved to an agreement?  Who exactly is in charge here?  This PR also fixes deterministic r.now problems addressed by PR #6191, in fact it includes its commit.

Somebody else has to merge this PR, I'm not going to.  If you want to use `r.write_timestamp()` instead, an implementation for that is two commits up, as of "Removed duplicate write hook checking logic."

This PR should be merged or rebased onto v2.4.x and then cherry-picked onto next.

### CODE REVIEW

File: drivers/javascript/ast.coffee [hunk 1]
```diff
@@ -1288,6 +1288,8 @@ rethinkdb.tableList = (args...) -> new TableList {}, args...
 
 rethinkdb.grant = (args...) -> new Grant {}, args...
 
+rethinkdb.writeTimestamp = (args...) -> new WriteTimestamp {}, args...
+
```
Comment (marshall007) [inline drivers/javascript/ast.coffee]: Need to revert these driver changes.

Comment (marshall007) [inline drivers/javascript/ast.coffee]: @srh need to revert these driver changes.

File: drivers/python/rethinkdb/query.py [hunk 2]
```diff
@@ -95,6 +96,10 @@ def grant(*args, **kwargs):
     return ast.GrantTL(*args, **kwargs)
 
 
+def write_timestamp(*args):
+    return ast.WriteTimestamp(*args)
+
+
```
Comment (marshall007) [inline drivers/python/rethinkdb/query.py]: Revert

**Comment by srh** [comment]
Agh, just dropped the WRITE_TIMESTAMP stuff from the drivers...

**Comment by srh** [comment]
Okay those are reverted...

File: src/pprint/js_pprint.cc [hunk 3]
```diff
@@ -684,6 +684,7 @@ counted_t<const document_t> render_as_javascript(const ql::raw_term_t &t) {
 // - `should_continue_string` should add the new Term to the list if
 //   it should be used in a string-of-dotted-expressions context.  So
 //   like `r.foo(1).bar(2).baz(4)` instead of `r.eq(1, 2)`.
+//   ^^ What does this mean??
```
Comment (marshall007) [inline src/pprint/js_pprint.cc]: I think it's referring to cases where you'd naively pretty-print `r.eq(var_0, 1)`, but you'd prefer `var_0.eq(1)`. The inverse example would be that you don't want `r.expr('foo').db()`, it should be `r.db('foo')`.

Comment (marshall007) [inline src/pprint/js_pprint.cc]: Actually I'm wrong about the `r.eq()` example. It appears the pretty-printer always prefers `r.eq(args...)`, but I think the `.db()` one holds.

Comment (srh) [inline src/pprint/js_pprint.cc]: Hopefully that comment will provide emotional support to other people that are confused.

**Comment by danielmewes** [comment]
API and design sound great to me. I've only done a quick pass over the code so far. Will review more closely in the next days.

Can we add a test case for the timestamp API?

**Comment by srh** [comment]
We can, if by we you mean somebody other than me, because I'm taking the next week off from working on RethinkDB.

**Comment by marshall007** [comment]
@srh yea I'll take care of the polyglot tests. Should have something up later tonight.

**Comment by srh** [comment]
Okay, that's great news.  I'm basically iPad-only for the next week so consider me to be mildly incapacitated.

**Comment by srh** [comment]
Your changes look good to me @marshall007.

**Comment by danielmewes** [comment]
Code looks good.
Thanks for making the `read_all_rows_filtered` thing and `deterministic_t` cleaner.

I say :shipit: 

Can you merge it or do you want me (or someone else) to do that?

**Comment by srh** [comment]
I can't merge it, I needed to give an external software access to my Github account and all my orgs, so I removed myself from the RethinkDB org.

**Comment by danielmewes** [comment]
Merged into v2.4.x, and cherry-picked into `next` as well.

**Comment by srh** [comment]
Thanks Daniel.

**Comment by robertjpayne** [comment]
Just running the polygot tests on v2.4.x branch is this still a valid test:

```
desc: Test that r.now() is allowed before changes() and after.
table_variable_name: tbl
tests:

    - py: tbl.changes().merge({'a': r.now()})
      ot: err('ReqlQueryLogicError')

    - py: tbl.merge({'a': r.now()}).changes()
```

It seems to fail the test because it works fine…

**Comment by marshall007** [comment]
@robertjpayne that test was added as part of this PR when the behavior of `r.now()` in the context of changefeeds was still somewhat undefined, but that was prior to https://github.com/rethinkdb/rethinkdb/pull/6396/commits/e6c15e48f8945987b4c6afe48ac9333c1ac512ba, which made `r.now()` deterministic across batches.

I think that failing test should now be removed.

**Pull request merged**